### ISSUE TITLE
Unable to update

### ISSUE THREAD

**Issue by danielo515** [dchester/epilogue#167]
Hello,

You may want to label this as question, because it is.

I am having problems updating certain elements using epilogue. If I edit the same elements using any sql editor (such SQLite studio) I can edit the fields that I want to edit, but if I try to do the same with epilogue I'm only able to update one field (the only that is not PK nor ForeignKey). It is not an issue of other elements using this element FK because it is not being used at the moment, and in that case the SQLite Studio throws an error too.

What it is confusing me the most is that epilogue is not trhowing any error, it just returns the original object and a 200 status.

Here is the model:

``` js
module.exports = function(sequelize, DataTypes) {
    return sequelize.define('DEVICE_MODEL', {
        DeviceModelName: {
            type: DataTypes.TEXT,
            allowNull: false,
            defaultValue: undefined,
            primaryKey: true
        },
        DeviceType: {
            type: DataTypes.TEXT,
            allowNull: false,
            defaultValue: undefined,
            primaryKey: true,
            references: {
                model: 'DEVICE_SUBTYPE',
                key: 'DeviceType'
            }
        },
        DeviceSubtype: { // change this if you gets back to the original table!
            type: DataTypes.TEXT,
            allowNull: false,
            defaultValue: undefined,
            primaryKey: true,
            references: {
                model: 'DEVICE_SUBTYPE',
                key: 'DeviceSubtypeName'
            }
        },
        DeviceManufacturer: {
            type: DataTypes.TEXT,
            allowNull: false,
            defaultValue: undefined,
            primaryKey: true,
            references: {
                model: 'DEVICE_MANUFACTURER',
                key: 'DeviceManufacturerName'
            }
        },
        DeviceUrlPicture: {
            type: DataTypes.TEXT,
            allowNull: true, // change if you rollback
            defaultValue: undefined
        }
    }, {
        tableName: 'DEVICE_MODEL'
    });
};
```

**Comment by danielo515** [comment]
Ok,

Sorry for opening this issue, seems to be a problem of sequelize, not epilogue:

https://github.com/sequelize/sequelize/issues/5827

Sorry again

**Comment by danielo515** [comment]
Ok,

Accroding to sequelize staff it can be easily avoided:

> It's a safety check, you generally don't want user provided values to edit the primary key.
> You can use raw: true to skip the check

The question is, how can I set raw to true on epilogue?

**Comment by mbroadst** [comment]
@danielo515 most (if not) all of the controllers support an `options` property on the context which they pass to the relevant sequelize function (e.g. https://github.com/dchester/epilogue/blob/master/lib/Controllers/read.js#L50 which is set here: https://github.com/dchester/epilogue/blob/master/lib/Controllers/read.js#L20). 

In order to pass `raw: true` to those options, you would need to create a milestone before the relevant action and set `context.options = { raw: true }`, in which case it will be included in the sequelize operation.

**Comment by danielo515** [comment]
Hello @mbroadst ,

Thank you for your help. I'm trying to figure how to install such millestone, assuming you mean an epilogue milestone. I'm trying it this way:

Declaring the hook:

```
module.exports.updateRawOption = {
    update: {
        start: {
            before: function (req, res, context) {
                console.log('Update hook called');
                context.options = { raw: true };
                return context.continue();
            }
        }
    }
}
```

Using the hook:

```
    // Create REST resource
    var modelResource=epilogue.resource({
        model: models.DEVICE_MODEL,
        endpoints: ['/devices/models', '/devices/models/:DeviceModelName']
    });
modelResource.use(hooks.updateRawOption); // this is an epilogue millestone hook
```

But the milestone is not being executed.
What I'm doing wrong?

**Comment by mbroadst** [comment]
@danielo515 this is because there is no defined `start` milestone for update. Try using `action` instead of `before` in your definition above. You would use `before` for `write` as an example here, since the update controller defines a `write` milestone already

**Comment by danielo515** [comment]
Hello @mbroadst 
Thank you for your help. I was trying changing the milestone definition to what you said (what a coincidence). By the way, documentation about which milestones are available needs an update. Now I'm getting some weird errors at bluebird, and module.js when I start the app, one of them fires on `fireDomevent` function,but apart from that seems like everything is working, the milestone is being called, but the options are not being taken into account by sequelize. Maybe options is not the proper property to set? This is how my definition looks now:

```javascript
module.exports.updateRawOption = {
    update: {
        write: {
            before: function (req, res, context) {
                console.log('Update hook called');
                context.options = { raw: true };
                return context.continue();
            }
        }
    }
}
```

**Comment by mbroadst** [comment]
@danielo515 it does indeed look like `options` are not taken into account in the Update controller: https://github.com/dchester/epilogue/blob/master/lib/Controllers/update.js#L22. 

Having said that, it doesn't seem like Sequelize even supports `raw: true` for saving an instance: http://docs.sequelizejs.com/en/latest/api/instance/#saveoptions-promisethiserrorsvalidationerror.

Can you provide an example of exactly what sequelize code would solve your problem (if epilogue was not involved)?

**Comment by mbroadst** [comment]
Scratch that, you want to pass raw into the initial fetch of the data, as you can see here Update uses Read's `fetch`: https://github.com/dchester/epilogue/blob/master/lib/Controllers/update.js#L20. 

That means that you'll want to create your milestone on `{ update: { fetch: { before: ...`

**Comment by danielo515** [comment]
Hello @mbroadst 

I'll try to give it a try to your last suggestion. Before, I want to say that I can see that the options are properly being appended to the instance object (which I think is the sequelize instance) as you can see on this screenshot:

![epilogue_](https://cloud.githubusercontent.com/assets/2270425/15761680/d164b224-2919-11e6-9aca-089269f35727.JPG)

**Comment by mbroadst** [comment]
@danielo515 yes but the problem is that you want `raw` to be true when you fetch the instance, and that's applying it after its already been fetched

**Comment by danielo515** [comment]
Here is the piece of sequelize code that I'm trying to achieve:

```javascript
// example model definition
var User = sequelize.define('users', {
  id: {
    type: DataTypes.STRING,
    unique: true,
    primaryKey: true,
    field: 'id'
  }
)};


User.findById(oldId).then(function(user) {
 user.set('id', 42, { raw: true });
 user.changed('id', true);
 user.save(); 
});
```

**Comment by mbroadst** [comment]
@danielo515 hm this might be a little more difficult for you then, as we still use `Instance.setAttributes` rather than `Instance.set`. `setAttributes` doesn't take an `options` object, so you might need to modify epilogue directly and work towards a PR to support this

**Comment by mbroadst** [comment]
@danielo515 actually, looking at their code it might be as easy as converting our use of `setAttributes` directly to `set`. Try changing this line: https://github.com/dchester/epilogue/blob/master/lib/Controllers/update.js#L48

to:
```
instance.set(context.attributes, context.options);
```

**Comment by danielo515** [comment]
I feel like we are getting closer :smile: 

Using the set, instead of set attributes makes me closer, but still some things missimng.
Here is the set code:
https://github.com/sequelize/sequelize/blob/d25456441a1fdd92216c37b06a802f9d34b44e56/lib/model.js#L2917

It expects an options object, the options object , where the raw option shoud be set, as it is checked here:
https://github.com/sequelize/sequelize/blob/d25456441a1fdd92216c37b06a802f9d34b44e56/lib/model.js#L2933
or here
https://github.com/sequelize/sequelize/blob/d25456441a1fdd92216c37b06a802f9d34b44e56/lib/model.js#L2933

Sadly the call you to set you have suggested is missing the options object. There is another options object with a dolar sign which has the option set, but this object is not being used

As you can see on the image below, epilogue is setting the option into the `$options` of the instance, but what we actually need is to pass that options to the set function
![sequelize](https://cloud.githubusercontent.com/assets/2270425/15772409/369a48a0-2971-11e6-998f-446b92849467.JPG)

**Comment by danielo515** [comment]
I'm also disturbing on sequelize repository here:
https://github.com/sequelize/sequelize/issues/5827#issuecomment-223522684

They have recomended me to use bulk update instead of save, which seems totally impossible with epilogue in an easy way (or maybe in any way at all).
